### PR TITLE
Limit empty lines in functions or methods body

### LINKED ISSUE #2486: Empty lines in method body
# Rule request

Empty lines in the method body are bad. I propose prohibit their use or limit it count in method body

## Thesis

I received a PR with the same code

```python
    def to_representation(self, data):
        result = None

        self.get_person(data)

        if self.person and self.person.use_other_veh:

            result = super().to_representation(data)

        return result
```

## Reasoning

I think, if method have few empty lines, it not holistic, and it makes sense to divide the method into several. Yegor Bugaenko's [post](https://www.yegor256.com/2014/11/03/empty-line-code-smell.html) about empty lines

### PR DESCRIPTION
**Pull request by blablatdinov**
# I have made things!

## Checklist

- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [x] I have created at least one test case for the changes I have made
- [x] I have updated the documentation for the changes I have made
- [x] I have added my changes to the `CHANGELOG.md`

## Related issues

Format is:
- Closes #2486

🙏 Please, if you or your company is finding wemake-python-styleguide valuable, help us sustain the project by sponsoring it transparently on https://opencollective.com/wemake-python-styleguide. As a thank you, your profile/company logo will be added to our main README which receives hundreds of unique visitors per day.

### CODE REVIEW

**Review by sobolevn — COMMENTED** [review]
Please, add configuration and default ration (I think that default amount of empty lines should be 1-for-2 non-empty lines).

**Comment by blablatdinov** [comment]
@sobolevn I don't agree or don't correctly understand you proportions. In this proportions linter don't reject next code:

```python
def func(a, b):
    first_expression(a)

    second_expression(b)
```

Or you implied 2 line with expressions without gap between them?

**Comment by sobolevn** [comment]
We can add a lower bound of 0 empty lines for 2 lines. That's not a problem.

**Comment by blablatdinov** [comment]
@sobolevn I'm pushed commit with code for configurating this rule. Please, check it

**Comment by sobolevn** [comment]
Can you please rebase your work? I've fixed new poetry@1.2.0 build

**Comment by blablatdinov** [comment]
Yes, of course. My branch rebased, now

**Comment by blablatdinov** [comment]
Sorry, my bad. I connect my visitor and found a problem: `ast` module ignore comments, so my algorithm count it like empty line. I've been looked solution of this problem on Google and I can't find it (in stackoverflow recomends using `tokenize` module). Do you have recomendation for resolve this problem ?

File: tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        foo()
+
+        bar()
+
+        baz()
+
+        lighter()
+"""
+
+class_with_allow_method = """
```
Comment (skarzi) [inline tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py:24]: NIP: what do you think about using `valid` instead of `allow`?
`class_with_valid_method` sounds better

File: tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+def func():
+    foo()
+    if name == 'Moonflower':
+        print('Love')
+    baz()
+"""
+
+
+allow_function_with_comments = """
+def log_costomer_info(customer):
```
Comment (skarzi) [inline tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py:59]:
```suggestion
def log_customer_info(customer):
```

File: wemake_python_styleguide/options/config.py [hunk 3]
```diff
@@ -267,6 +270,11 @@ class Configuration(object):
             type='string',
             comma_separated_list=True,
         ),
+        _Option(
+            '--exps-for-one-empty-line',
+            defaults.EXPS_FOR_ONE_EMPTY_LINE,
+            'Count of expressions for one empty line if function body.',
```
Comment (skarzi) [inline wemake_python_styleguide/options/config.py:276]:
```suggestion
            'Count of expressions for one empty line in a function body.',
```

File: wemake_python_styleguide/options/defaults.py [hunk 4]
```diff
@@ -46,6 +46,9 @@
 #: Violation codes that are forbidden to use.
 FORBIDDEN_INLINE_IGNORE: Final = ()
 
+#: Count of available expressions for one empty line if function or method body.
```
Comment (skarzi) [inline wemake_python_styleguide/options/defaults.py:49]:
```suggestion
#: Count of available expressions for one empty line in function or method body.
```

File: wemake_python_styleguide/violations/best_practices.py [hunk 5]
```diff
@@ -2770,3 +2772,48 @@ class GettingElementByUnpackingViolation(ASTViolation):
...
+
+
+@final
+class WrongEmptyLinesCountViolation(ASTViolation):
+    """
+    Limit empty lines in functions or methods body.
+
+    Reasoning:
+        Functions or methods have few empty lines, it not holistic,
+        and it makes sense to divide the method into several.
```
Comment (skarzi) [inline wemake_python_styleguide/violations/best_practices.py:2784]:
```suggestion
        It's not holistic to have functions or methods that contain many empty lines,
        and it makes sense to divide the method into several ones.
```

File: wemake_python_styleguide/violations/best_practices.py [hunk 6]
```diff
@@ -2770,3 +2772,48 @@ class GettingElementByUnpackingViolation(ASTViolation):
...
+        This rule is configurable with ``--exps-for-one-empty-line``.
+        Default:
+        :str:`wemake_python_styleguide.options.defaults.EXPS_FOR_ONE_EMPTY_LINE`
+
+    .. versionadded:: 0.17.0
+
+    """
+
+    error_template = (
+        'Found wrong empty lines count: {0}'
```
Comment (skarzi) [inline wemake_python_styleguide/violations/best_practices.py:2817]:
```suggestion
        'Found too many empty lines: {0}'
```

File: wemake_python_styleguide/visitors/ast/function_empty_lines.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        lines_range = set(range(start_line, node.end_lineno))
+        lines_without_expressions = self._lines_without_expressions(
+            copy(lines_range), node,
+        )
+        empty_lines_count = len(lines_without_expressions)
+        available_empty_lines = self._available_empty_lines(
+            len(lines_range - lines_without_expressions),
+        )
+        if not empty_lines_count:
+            return
```
Comment (skarzi) [inline wemake_python_styleguide/visitors/ast/function_empty_lines.py:36]: NIP: you can define `if` condition before calculating `available_empty_lines` to make code more efficient in cases there is no empty lines:

```suggestion
        if not empty_lines_count:
            return
        available_empty_lines = self._available_empty_lines(
            len(lines_range - lines_without_expressions),
        )
```

File: wemake_python_styleguide/visitors/ast/function_empty_lines.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    def _lines_without_expressions(
+        self,
+        lines_range: set[int],
+        node,
+    ) -> set[int]:
+        for subnode in ast.walk(node):
+            if isinstance(subnode, ast.Constant) and subnode.end_lineno:
+                lines_range = lines_range - set(range(
+                    subnode.lineno, subnode.end_lineno + 1,
+                ))
```
Comment (skarzi) [inline wemake_python_styleguide/visitors/ast/function_empty_lines.py:56]:
```suggestion
                lines_range -= set(range(
                    subnode.lineno, subnode.end_lineno + 1,
                ))
```

File: wemake_python_styleguide/visitors/ast/function_empty_lines.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    """Restricts empty lines in function or method body."""
+
+    def visit_any_function(self, node: types.AnyFunctionDef) -> None:
+        """Find empty lines count."""
+        start_line = node.lineno
+        if not start_line or not node.end_lineno:
+            return
+        lines_range = set(range(start_line, node.end_lineno))
+        lines_without_expressions = self._lines_without_expressions(
+            copy(lines_range), node,
```
Comment (skarzi) [inline wemake_python_styleguide/visitors/ast/function_empty_lines.py:29]: NIP: `set` has `copy` method that can be used to create shallow copy of a `set` instance:

```suggestion
            lines_range.copy(), node,
```

File: tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    assert_errors,
+    assert_error_text,
+):
+    """Testing available cases."""
+    tree = parse_ast_tree(class_with_wrong_method)
+
+    visitor = WrongEmptyLinesCountVisitor(default_options, tree=tree)
+    visitor.run()
+
+    assert_errors(visitor, [WrongEmptyLinesCountViolation])
```
Comment (skarzi) [inline tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py:108]: Is this assertion correct? It seems like in this test `input_` has only a valid number of empty lines

File: tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+wrong_function = """
+def func():
+    foo()
+
+    a = 1 + 4
+
+    baz()
+"""
+
+wrong_function_with_loop = """
```
Comment (skarzi) [inline tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py:42]: What about edge cases like below ones?

```python
def foo():
    return (
        pekaboo['attr'] == 'dummy-value'
        and bar() == 'Knights that say Ni!'


        or (
            'attr2' in pekaboo
            and not some_object.attribute.method()
        )
    )
```

```python
pekaboo(
    foo=123,



    bar='abc',
)
```

Can they be handled by the `ast`?
Probably this violation should be implemented by using `BaseTokenVisitor`

File: tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py (hunk 1 above)
Comment (blablatdinov) [reply tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py:24]: Yes, I'm agree. My english level is not high, so I'm happy with any edits

File: tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py (hunk 11 above)
Comment (blablatdinov) [reply tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py:42]: Yeh, I plan to change visitor to `BaseTokenVisitor` because `ast` ignore comments. I check these cases after changing the visitor, thank you

File: tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py (hunk 10 above)
Comment (blablatdinov) [reply tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py:108]: I forgot to change `parse_ast_tree` after parameterizing the tests, thanks

File: wemake_python_styleguide/visitors/ast/function_empty_lines.py (hunk 9 above)
Comment (blablatdinov) [reply wemake_python_styleguide/visitors/ast/function_empty_lines.py:29]: Didn't know about this method

File: wemake_python_styleguide/visitors/ast/function_empty_lines.py (hunk 7 above)
Comment (blablatdinov) [reply wemake_python_styleguide/visitors/ast/function_empty_lines.py:36]: Yes, I agree

**Comment by blablatdinov** [comment]
I'm change visitor to `BaseTokenVisitor`. Can you review it?

File: tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+def func():
+    foo()
+    if name == 'Moonflower':
+        print('Love')
+
+    baz()
+"""
+
+
+allow_function_with_comments = """
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py:65]: Please, add this case:

```python
def test_func():
   """
   Its docstring

   has many new lines
   
   but this is 

   totally fine
 
   we don't raise a violation for this
   """
   return
```

File: tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+    wrong_function_with_loop,
+])
+def test_wrong(
+    input_,
+    default_options,
+    assert_errors,
+    parse_tokens,
+):
+    """Testing wrong cases."""
+    file_tokens = parse_tokens(input_)
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py:88]: Don't forget to use `mode` fixture to test `async def` as well.

File: tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+
+allow_function_with_comments = """
+def log_customer_info(customer):
+    # printing customer name
+    print(customer.name)
+    # printing customer phone
+    print(customer.phone)
+    # printing customer company
+    print(customer.company)
+"""
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py:73]: We also need a test case like this:

```python
def test_func():
   # This function 
   # 
   # has lots 
   #
   # of empty 
   #
   # lines 
   #
   # in comments
   return 0
```

Comment (sobolevn) [reply tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py:73]: A mix of comments and docstring is also a good idea 👍

**Comment by blablatdinov** [comment]
Thanks for comments. I'm add test cases with comments and docstrings

File: wemake_python_styleguide/violations/best_practices.py [hunk 15]
```diff
@@ -2770,3 +2772,49 @@ class GettingElementByUnpackingViolation(ASTViolation):
...
+        Default:
+        :str:`wemake_python_styleguide.options.defaults.EXPS_FOR_ONE_EMPTY_LINE`
+
+    .. versionadded:: 0.17.0
+
+    """
+
+    error_template = (
+        'Found too many empty lines: {0}'
+    )
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/best_practices.py:2819]:
```suggestion
    error_template = 'Found too many empty lines in `def`: {0}'
```

File: wemake_python_styleguide/violations/best_practices.py [hunk 16]
```diff
@@ -2770,3 +2772,49 @@ class GettingElementByUnpackingViolation(ASTViolation):
...
+    """
+    Limit empty lines in functions or methods body.
+
+    Reasoning:
+        It's not holistic to have functions or methods that contain many
+        empty lines, and it makes sense to divide the method into several
+        ones.
+
+    Solution:
+        Limit count of empty lines of the function or method body
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/best_practices.py:2788]:
```suggestion
    Solution:
        Limit count of empty lines of the function or method body
        By default, we allow 1 empty line for 2 non-empty lines.
```

File: tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+def test_zero_option(
+    parse_tokens,
+    default_options,
+    assert_errors,
+    options,
+    mode,
+):
+    """Test zero configuration."""
+    file_tokens = parse_tokens(mode(allow_function))
+    visitor = WrongEmptyLinesCountVisitor(
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_classes/test_wrong_empty_lines_count.py:177]: Let's also add a test that function without any empty lines do pass with `exps_for_one_empty_line=0`

**Review by sobolevn — COMMENTED** [review]
Awesome! This is the last piece of feedback that I have.

**Comment by blablatdinov** [comment]
Thanks for your suggested changes! I'm add test with `exps_for_one_empty_line=0` for method without empty lines

**Review by sobolevn — APPROVED** [review]
Thank you for a great idea and solid implementation!
👏

**Comment by blablatdinov** [comment]
Thank you too, this is my first contribution to open source with code! The "quality wall" in this repository is very high and strong, respect to the developers who build it. 🎉

I bought the course "typed python" from @f213 and @sobolevn, we'll meet there 😉
